### ISSUE TITLE
UI Spoofing

### ISSUE THREAD

**Issue by indolering** [mailvelope/mailvelope#293]
My apologies if this the wrong venue to begin a discussion, google didn't turn up a mailing list and I didn't see any mention of one on the website.

I was reviewing your recent ["pen test"](https://cure53.de/pentest-report_mailvelope.pdf) and I disagree with the conclusions they drew about the level of mitation provided by your watermarking.

New installations of Mailvelope generate a new watermark.  As long as an attacker can detect Mailvelope's content injections, they can fingerprint the existing browser and wait until the user visits the site with a new browser and insert their own spoofed UI.  This can be mitigated by asking users to choose their own security icons, so that the icons are consistent across installations.

However, I recently performed a literature review of security icons and I believe that their efficacy is very low.  Nearly every research study on SSL security indicators finds a near 100% success rate: users just don't notice the lack of the icon.  However, studies that dedicating the entire browser browser chrome to conveying security information found only modest improvements.

More to the point, [The Emperor’s New Security Indicators](http://commerce.net/wp-content/uploads/2012/04/The%20Emperors_New_Security_Indicators.pdf) found that 36% of users logged into their personal bank accounts even when the site authentication images were removed.

The fundamental issue is that human cognition has limits: we cannot process unlimited amounts of information.  Indeed, human cognition is largely serial and we can only perform multitasking if we can habituate the background task.  Learning to drive a car initially requires conscious effort for each task, such as turning on the blinker, slowing down, and turning the wheel of the car.  Users will not habituate behaviors to icons over time because, unlike a red light, ignoring them does not impede the task at hand.

I still think Mailvelope is useful for users that do not face dedicated attackers.  However, I do not believe that Mailvelope should be used by high-risk groups.

**Comment by indolering** [comment]
Correction, The Emperor’s New Security Indicators study (which removed site authentication images and asked people to login to their bank account) had a 92% success rate.

**Comment by leonard84** [comment]
Interesting study, do you have any suggestions as to alternatives or improvements to prevent UI spoofing?

**Comment by indolering** [comment]
@leonard84 Nope, I'm writing a blog post about it now.  As far as I can tell, it's a head shot.

**Comment by taoeffect** [comment]
@leonard84 Here's the blog post he wrote: http://www.indolering.com/e2e-web-crypto

FYI, I recommend not engaging with Zach. His blog posts are like 20% accurate statements, 60% misleading statements, and 20% outright false statements.

When you point this out to him, he will not correct his blog posts, he will update them with more false or misleading statements. At least that's my experience so far with him.

So I've basically given up on having honest conversations with him.

**Comment by indolering** [comment]
@taoeffect I already removed okTurtles from that post so that I wouldn't have to deal with you.

I haven't posted that anywhere (including) because it wasn't done, I am getting feedback from others.  I've already thought of one possibility for fixing it for okTurtles.

@leonard84 I think the only way to mitigate against spoofing attacks is if you are signing all of the messages by default *and* if the recipient's software shows warnings when things are not signed.  The problem is that GPG/PGP is super un-opinionated and it doesn't do this by default.

okTurtles might have a better shot if they moved to OTR, because (AFAIK) most of those clients do validation by default.  I don't know that for sure, which is why I haven't pushed this anywhere.  OTTH, Facebook and Twitter would have total control of the UI on both ends.  So they could just fake the validation on both ends.

**Comment by indolering** [comment]
... and those warnings cannot be small.  They basically have to block the users workflow, think invalid certificate warnings on browsers.

**Comment by indolering** [comment]
Okay, I've pushed the final version of [the blogpost](http://www.indolering.com/e2e-web-crypto).

* New message needs to be outside of the website and in the browser chrome.
* I *think* automatically pasting into a new message is safe. 
* Hinting from the "to" address is probably safe.

I don't know how you can handle decryption without dropping UI elements into the native interface. I don't think you should  display decrypted contents in-line, as a service provider can drop a dummy reply button on it.

**Comment by indolering** [comment]
If you decide to continue with dropping UI elements into the native interface, you need to display warnings that it isn't safe against targeted attacks and shouldn't be used by high-risk groups that could motivate an attack (such as journalists and human rights workers).

**Comment by Manouchehri** [comment]
@indolering If your service provider is attempting to hijack/phish your browser, I think there's more serious issues that are beyond Mailvelope.

High-risk groups should probably not be using email at all, period. There's quite a few countries where [*using* PGP/encryption](http://www.cryptolaw.org/cls2.htm) is considered [an indication of] a crime. [Even the mail host](https://www.eff.org/deeplinks/2015/01/security-not-crime-unless-youre-anarchist) can be an issue in some places.

**Comment by indolering** [comment]
@Manouchehri I agree with you that the vast majority of users don't need worry about this particular attack.  And I'm not saying that Mailvelope doesn't provide value; my boss has started to use PGP for casual emails and I'm thinking of adding some lower-security PGP keys that I can use with Mailvelope.  

However, the use of PGP in conjunction with an email client provides *much* stronger protections against malicious service providers than what Mailvelope can achieve.  This distinction is not /at all/ obvious to users and the continued reliance on security mechanisms with such high failure rates, an uncorrected security audit, and a lack of strong disclosures only compounds the problem.  

Not only does it place *your* users at risk, but it perpetuates a security model that relies on a mechanism for protecting against malicious service providers that has an extremely high failure rate.  I posted here because when I raised these concerns with another project, they pointed to Mailvelope's security audit.  This external project made *very* strong claims about being "MITM Proof" and explicitly mentioned journalists and human rights workers in their marketing.  

I'm not happy about all this, I got into security focused usability research because I recognize that usability is a prerequisite to security.   Early on, I was really enthusiastic about add-ons that inject E2E into web interfaces.  But I found major flaws to this approach, flaws that I couldn't fix.  I'm just trying to disclose these security issues responsibly.

**Comment by Manouchehri** [comment]
No need to create extra master keys, you can simply create a subkey and use that. [I've written how to do this](https://wiki.archlinux.org/index.php/GnuPG#Exporting_subkey) on Arch, and lots of other people have similar guides as well.

What you're describing sounds more like phishing rather than a man in the middle attack. If somebody obtains a valid certificate (that also passes pinning) for `mail.google.com`, I think it's fair to assume that Google would be financially motivated to protect their >400 million regular Gmail user base, Mailvelope aside. Stolen certificates have happened in the past, which is why Google finally had enough and introduced pinning.

sslstrip is difficult to apply in a real world attack now due to HSTS. [RFC-6797](https://tools.ietf.org/html/rfc6797) covered this a few years ago and all modern browsers follow it. A more effective strategy would be obtaining a SSL certification for a similar domain, such as `googie.com`. However, Mailvelope would not be run since it no longer matches the host wildcard.

You could use a combination of Mailvelope spoofing and a similar looking domain to achieve what you've mentioned (no need to strip out HTTPS, since a valid SSL cert would be automatically issued if requested). I could be wrong about this, but I believe all browsers still don't allow a webpage to create a popup without a URL bar. If they didn't, you could easily spoof the appearance of a system update and do a lot worse.

![Application vs. website](https://cloud.githubusercontent.com/assets/7232674/9696629/e12046dc-5341-11e5-9bf0-47cb335a39bb.png)

Also, you'll notice that clicking inside a non-spoofed window results in the word "Ok" to be flashed on top of the icon (in the browser bar). To summarize, the following requirements would need to be met for a successful attack:

1. A stolen SSL certificate from the mail provider.
2. The user not noticing an incorrect Mailvelope watermark.
3. The user not noticing the addition of a URL bar in "Mailvelope".
4. The user not noticing the lack of response from the Mailvelope icon.

*OR*

1. The user going to a phishing site.
2. The user entering their password into the previously mentioned site. (At this point an attacker has a valid password and could send phishing emails to the user through their own account, bypassing any spam filter.)
3. The user not noticing an incorrect Mailvelope watermark.
3. The user not noticing the addition of a URL bar in "Mailvelope".
4. The user not noticing the lack of response from the Mailvelope icon.

If none of those issues set a red flag off for the user, they're extremely likely to fall for any phishing attack. A dedicated mail client will not prevent them from running a file named `documents.pdf.exe` either.

Searching `site:mailvelope.com journalist OR activists OR "human rights"` on Google doesn't return any matches. Can you provide a link so it's easier for us to follow?
